### PR TITLE
Add ability to create any k8s resource from Flyte's resource controller

### PR DESCRIPTION
**Pull request by pradithya**
# TL;DR
Allow Flyte's resource controller to provision any k8s resource

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
Flyte is able to create kubernetes resource for each of Flyte's project via Flyte's resource controller. Flyte admin can customise what resource to be created by specifying k8s template file(s) such as the [example](https://github.com/flyteorg/flyteadmin/tree/master/sampleresourcetemplates). 

Currently, this feature is limited to provisioning k8s resource recognised by the controller's  decoder (and currently it is able to recognize standard k8s resources). It's the case because current implementation requires decoding the kubernetes manifest (yaml) into kubernetes object, thus the schema must be registered first. 
The PR remove the necessity of decoding the k8s manifest hence allowing any resource to be created (as long as it's recognized by the k8s api) 
In my case, this feature is useful for automating Flyte's project creation when it need to create a separate [Google Service Account](https://cloud.google.com/config-connector/docs/reference/resource-docs/iam/iamserviceaccount) via [config connector](https://cloud.google.com/config-connector/docs/overview) for each Flyte's project to be used by task execution pod when using any Google's offering (BQ, GCS, etc)

```
apiVersion: iam.cnrm.cloud.google.com/v1beta1
kind: IAMServiceAccount
metadata:
  name: {{ namespace }}-gsa
  namespace: {{ namespace }}
spec:
  displayName: Service account for {{ namespace }}
```

### CODE REVIEW

**Comment by kumare3** [comment]
is there a way to just create the raw json, without decoding the object at all? IIUC, the problem is that the scheme is not standard and hence unknown.
Is it possible to use something like [k8s unstructured client](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured) ?

**Comment by pradithya** [comment]
> is there a way to just create the raw json, without decoding the object at all? IIUC, the problem is that the scheme is not standard and hence unknown. Is it possible to use something like [k8s unstructured client](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured) ?

I think it is possible to make it similar to `kubectl`. Let me get better understanding why current approach requires decoding.

**Comment by pradithya** [comment]
While smoke testing, I managed to create `IAMServiceAccount` resource

```
apiVersion: iam.cnrm.cloud.google.com/v1beta1
kind: IAMServiceAccount
metadata:
  annotations:
    cnrm.cloud.google.com/management-conflict-prevention-policy: none
    cnrm.cloud.google.com/project-id: flyteexamples
    cnrm.cloud.google.com/state-into-spec: merge
  creationTimestamp: "2021-11-19T06:06:16Z"
  finalizers:
  - cnrm.cloud.google.com/finalizer
  - cnrm.cloud.google.com/deletion-defender
  generation: 1
  name: flyteexamples-gsa
  namespace: flyteexamples
  resourceVersion: "51426499"
  selfLink: /apis/iam.cnrm.cloud.google.com/v1beta1/namespaces/flyteexamples/iamserviceaccounts/flyteexamples-gsa
  uid: cf2d53a1-08d9-495f-b75d-4d323e36aaa5
```

But in subsequent sync it causes the resource controller to crashes with following errors. Any idea?

```
{"json":{"src":"controller.go:522"},"level":"warning","msg":"Failed to create cluster resources for namespace [flyteexamples] with err: iamserviceaccounts.iam.cnrm.cloud.google.com \"flyteexamples-gsa\" is invalid: metadata.resourceVersion: Invalid value: 0x0: must be specified for an update","ts":"2021-11-19T06:44:12Z"}
{"json":{"src":"clusterresource.go:112"},"level":"fatal","msg":"Failed to sync cluster resources [iamserviceaccounts.iam.cnrm.cloud.google.com \"my-project-gsa\" is invalid: metadata.resourceVersion: Invalid value: 0x0: must be specified for an update, iamserviceaccounts.iam.cnrm.cloud.google.com \"flytetester-gsa\" is invalid: metadata.resourceVersion: Invalid value: 0x0: must be specified for an update, iamserviceaccounts.iam.cnrm.cloud.google.com \"flytesnacks-gsa\" is invalid: metadata.resourceVersion: Invalid value: 0x0: must be specified for an update, iamserviceaccounts.iam.cnrm.cloud.google.com \"flyteexamples-gsa\" is invalid: metadata.resourceVersion: Invalid value: 0x0: must be specified for an update]","ts":"2021-11-19T06:44:12Z"}
```

**Comment by katrogan** [comment]
does updating a Google Service Account require additional parameters not in the create? are you able to test updating without the cluster resource controller just by applying yaml?

**Comment by pradithya** [comment]
> does updating a Google Service Account require additional parameters not in the create? are you able to test updating without the cluster resource controller just by applying yaml?

Yes,  applying them directly using kubectl works fine. 
I tried to ensure the update happens using patch (by adding the resource kind into [this mapping](https://github.com/flyteorg/flyteadmin/blob/1373fa4e8eb83893dbd78e6e5de132e5991a524e/pkg/clusterresource/controller.go#L120), but it doesn't seem to work too, likely due to https://github.com/kubernetes/website/pull/10994

**Comment by kumare3** [comment]
@pradithya we should never call and update. We should simply call create. Update needs the resource version of the current object. So perform an update you have to fetch, merge and then update. 

https://github.com/kubernetes/apimachinery/blob/7149480564b7d40046f19040efa20171b49059ae/pkg/apis/meta/v1/types.go#L78

**Comment by kumare3** [comment]
@katrogan why do we update? Is
It for secrets

**Comment by pradithya** [comment]
Updated the PR to allow any unknown resource to be created by cluster resource controller.

The issue with update still persists though, I'll need your advise on that @kumare3 @katrogan

**Comment by katrogan** [comment]
@kumare3 we actually use update by default, k8s handles reconciling with the existing resource. we only use strategic merge patch for serviceaccounts because update was actually resulting in creating a new one each time, which quickly overwhelms k8s api server when the number of resources grows large

@pradithya sorry this is extra work, but it looks JSONPatch and JSON merge patch [work](https://github.com/kubernetes/kubernetes/issues/58414#issuecomment-383148632) perhaps you could implement that and we could similarly special-case handle JSONPatch update-types (e.g. CRDs). since i believe the current default behavior to call Update() actually ends up calling http [PUT](https://github.com/kubernetes/client-go/blob/master/dynamic/simple.go#L161)

**Comment by pradithya** [comment]
Latest implementation uses Patch for all resources with following strategy:
1. Strategic merge patch for known resource
2. JSON merge patch for custom resource
The implementation heavily follows `kubectl apply` behaviour to ensure more use cases.
It will also skip the patch if the controller doesn't detect any modification.

Have been testing it in my dev cluster and so far so good, it managed to create new resource (no change in this code path) and sync existing resource for both known and custom resource.

```
{"json":{"src":"controller.go:343"},"level":"debug","msg":"Attempting to create resource [ServiceAccount] in cluster [] for namespace [flyteexamples]","ts":"2021-11-23T16:07:10Z"}
{"json":{"src":"controller.go:351"},"level":"debug","msg":"Type [ServiceAccount] in namespace [flyteexamples] already exists - attempting update instead","ts":"2021-11-23T16:07:10Z"}
{"json":{"src":"controller.go:383"},"level":"info","msg":"Resource [ServiceAccount] in namespace [flyteexamples] is not modified","ts":"2021-11-23T16:07:10Z"}
{"json":{"src":"controller.go:481"},"level":"debug","msg":"successfully read template config file [ab_project_resource_quota.yaml]","ts":"2021-11-23T16:07:10Z"}
{"json":{"src":"controller.go:343"},"level":"debug","msg":"Attempting to create resource [ResourceQuota] in cluster [] for namespace [flyteexamples]","ts":"2021-11-23T16:07:10Z"}
{"json":{"src":"controller.go:351"},"level":"debug","msg":"Type [ResourceQuota] in namespace [flyteexamples] already exists - attempting update instead","ts":"2021-11-23T16:07:11Z"}
{"json":{"src":"controller.go:398"},"level":"debug","msg":"Successfully updated resource [ResourceQuota] in namespace [flyteexamples]","ts":"2021-11-23T16:07:11Z"}
{"json":{"src":"controller.go:481"},"level":"debug","msg":"successfully read template config file [abb_google_service_account.yaml]","ts":"2021-11-23T16:07:11Z"}
{"json":{"src":"controller.go:343"},"level":"debug","msg":"Attempting to create resource [IAMServiceAccount] in cluster [] for namespace [flyteexamples]","ts":"2021-11-23T16:07:11Z"}
{"json":{"src":"controller.go:351"},"level":"debug","msg":"Type [IAMServiceAccount] in namespace [flyteexamples] already exists - attempting update instead","ts":"2021-11-23T16:07:11Z"}
{"json":{"src":"controller.go:383"},"level":"info","msg":"Resource [IAMServiceAccount] in namespace [flyteexamples] is not modified","ts":"2021-11-23T16:07:11Z"}


```

**Comment by pradithya** [comment]
@katrogan @kumare3 It can be reviewed again

File: pkg/clusterresource/controller.go [hunk 1]
```diff
@@ -447,6 +499,69 @@ func (c *controller) createResourceFromTemplate(ctx context.Context, templateDir
...
+// for custom resource, json merge patch is used
+// heavily inspired by kubectl's patcher
+func (c *controller) createPatch(mapping *meta.RESTMapping, currentObj *unstructured.Unstructured, modified []byte, namespace string) ([]byte, types.PatchType, error) {
+	current, err := k8sruntime.Encode(unstructured.UnstructuredJSONScheme, currentObj)
+	if err != nil {
+		return nil, "", fmt.Errorf("Failed to encode [%+v] in namespace [%s] to json with err: %v",
+			currentObj.GetKind(), namespace, err)
+	}
+
+	original, err := getLastApplied(currentObj)
```
Comment (katrogan) [inline pkg/clusterresource/controller.go:501]: do you ever check that err != nil?

File: pkg/clusterresource/controller.go [hunk 2]
```diff
@@ -447,6 +499,69 @@ func (c *controller) createResourceFromTemplate(ctx context.Context, templateDir
...
+
+	var patch []byte
+	patchType := types.StrategicMergePatchType
+	obj, err := scheme.Scheme.New(mapping.GroupVersionKind)
+	switch {
+	case err == nil:
+		patchType = types.StrategicMergePatchType
```
Comment (katrogan) [inline pkg/clusterresource/controller.go:512]: nit: this is what you initialize it to already in line 516

File: pkg/clusterresource/controller.go [hunk 3]
```diff
@@ -447,6 +499,69 @@ func (c *controller) createResourceFromTemplate(ctx context.Context, templateDir
...
+		if patch == nil {
```
Comment (katrogan) [inline pkg/clusterresource/controller.go:513]: isn't patch always nil at this point?

File: pkg/clusterresource/controller.go [hunk 4]
```diff
@@ -447,6 +499,69 @@ func (c *controller) createResourceFromTemplate(ctx context.Context, templateDir
...
+					currentObj.GetKind(), namespace, err, original, modified, current)
+			}
+		}
+	case k8sruntime.IsNotRegisteredError(err):
+		patchType = types.MergePatchType
+		preconditions := []mergepatch.PreconditionFunc{mergepatch.RequireKeyUnchanged("apiVersion"),
+			mergepatch.RequireKeyUnchanged("kind"), mergepatch.RequireMetadataKeyUnchanged("name")}
+		patch, err = jsonmergepatch.CreateThreeWayJSONMergePatch(original, modified, current, preconditions...)
+		if err != nil {
+			return nil, "", fmt.Errorf("Failed to create 3 way merge patch for resource [%+v] in namespace [%s] with err: %v",
```
Comment (katrogan) [inline pkg/clusterresource/controller.go]: is this case still 3 way merge patch?

File: pkg/clusterresource/controller.go [hunk 5]
```diff
@@ -447,6 +499,69 @@ func (c *controller) createResourceFromTemplate(ctx context.Context, templateDir
...
+					currentObj.GetKind(), namespace, err)
+			}
+
+			patch, err = strategicpatch.CreateThreeWayMergePatch(original, modified, current, lookupPatchMeta, true)
+			if err != nil {
+				return nil, "", fmt.Errorf("Failed to create 3 way merge patch for resource [%+v] in namespace [%s] with err: %v\noriginal:\n%s\nmodified:\n%s\ncurrent:\n%s",
+					currentObj.GetKind(), namespace, err, original, modified, current)
+			}
+		}
+	case k8sruntime.IsNotRegisteredError(err):
```
Comment (katrogan) [inline pkg/clusterresource/controller.go:526]: what does not registered in this case mean?

File: pkg/clusterresource/controller.go [hunk 6]
```diff
@@ -447,6 +499,69 @@ func (c *controller) createResourceFromTemplate(ctx context.Context, templateDir
...
+		patchType = types.MergePatchType
+		preconditions := []mergepatch.PreconditionFunc{mergepatch.RequireKeyUnchanged("apiVersion"),
+			mergepatch.RequireKeyUnchanged("kind"), mergepatch.RequireMetadataKeyUnchanged("name")}
+		patch, err = jsonmergepatch.CreateThreeWayJSONMergePatch(original, modified, current, preconditions...)
+		if err != nil {
+			return nil, "", fmt.Errorf("Failed to create 3 way merge patch for resource [%+v] in namespace [%s] with err: %v",
+				currentObj.GetKind(), namespace, err)
+		}
+	case err != nil:
```
Comment (katrogan) [inline pkg/clusterresource/controller.go:535]: nit: just `default` here

File: pkg/clusterresource/controller.go [hunk 7]
```diff
@@ -447,6 +499,69 @@ func (c *controller) createResourceFromTemplate(ctx context.Context, templateDir
...
+		return nil, "", fmt.Errorf("Failed to create get instance of versioned object [%+v] in namespace [%s] with err: %v",
+			currentObj.GetKind(), namespace, err)
+
+	}
+
+	if string(patch) == "{}" {
```
Comment (katrogan) [inline pkg/clusterresource/controller.go:541]: nit, can we make "{}" a named constant?

File: pkg/clusterresource/controller.go [hunk 8]
```diff
@@ -447,6 +499,69 @@ func (c *controller) createResourceFromTemplate(ctx context.Context, templateDir
...
+		// not modified
+		return patch, patchType, nil
+	}
+
+	if currentObj.GetResourceVersion() != "" {
```
Comment (katrogan) [inline pkg/clusterresource/controller.go:546]: nit: `if len(currentObj.GetResourceVersion()) > 0 {`

File: pkg/clusterresource/controller.go [hunk 9]
```diff
@@ -346,38 +352,47 @@ func (c *controller) syncNamespace(ctx context.Context, project models.Project,
...
+					patch, patchType, err := c.createPatch(dynamicObj.mapping, currentObj, modified, namespace)
+					if err != nil {
+						c.metrics.TemplateUpdateErrors.Inc()
+						logger.Warningf(ctx, "Failed to create patch for resource [%+v] in namespace [%s] err: %v",
+							dynamicObj.obj.GetKind(), namespace, err)
+						collectedErrs = append(collectedErrs, err)
+						continue
+					}
+
+					if string(patch) == "{}" {
```
Comment (katrogan) [inline pkg/clusterresource/controller.go:370]: can we use a named constant for "{}"?

**Review by katrogan — COMMENTED** [review]
this is awesome! thank you for cleaning up this code and adding such thorough unit tests

just a few nits and questions but looks great

File: pkg/clusterresource/controller.go (hunk 3 above)
Comment (pradithya) [reply pkg/clusterresource/controller.go]: Yep, missed this, updated now

File: pkg/clusterresource/controller.go (hunk 4 above)
Comment (pradithya) [reply pkg/clusterresource/controller.go]: it should be json merge patch, updated the error message

File: pkg/clusterresource/controller.go (hunk 5 above)
Comment (pradithya) [reply pkg/clusterresource/controller.go:524]: It's resources that's not recognized by client-go's schema by default. It's usually CRDs

File: pkg/clusterresource/controller.go (hunk 6 above)
Comment (pradithya) [reply pkg/clusterresource/controller.go]: Updated

File: pkg/clusterresource/controller.go (hunk 7 above)
Comment (pradithya) [reply pkg/clusterresource/controller.go]: Updated

File: pkg/clusterresource/controller.go (hunk 8 above)
Comment (pradithya) [reply pkg/clusterresource/controller.go]: Updated

File: pkg/clusterresource/controller.go (hunk 9 above)
Comment (pradithya) [reply pkg/clusterresource/controller.go]: Updated

File: pkg/clusterresource/controller.go [hunk 10]
```diff
@@ -452,6 +411,147 @@ func (c *controller) syncNamespace(ctx context.Context, project models.Project,
...
+	err := json.Unmarshal(patch, &patchMap)
+	if err != nil {
+		return nil, err
+	}
+	u := unstructured.Unstructured{Object: patchMap}
+	a, err := meta.Accessor(&u)
+	if err != nil {
+		return nil, err
+	}
+	a.SetResourceVersion(rv)
```
Comment (kumare3) [inline pkg/clusterresource/controller.go:446]: wow this is a little dangerous right?

**Review by kumare3 — DISMISSED** [review]

**Comment by kumare3** [comment]
@pradithya i think you are missing a DCO

File: pkg/clusterresource/controller.go (hunk 10 above)
Comment (pradithya) [reply pkg/clusterresource/controller.go:446]: Based on the docs it seems ok as long as we don't modify it.

```
	// String that identifies the server's internal version of this object that
	// can be used by clients to determine when objects have changed.
	// Value must be treated as opaque by clients and passed unmodified back to the server.
	// Populated by the system.
	// Read-only.
	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#concurrency-control-and-consistency
	// +optional
	ResourceVersion string `json:"resourceVersion,omitempty" protobuf:"bytes,2,opt,name=resourceVersion"`

```

**Comment by pradithya** [comment]
Updated the DCO

**Review by katrogan — APPROVED** [review]
